### PR TITLE
Make pointer_structural_match normal and warn

### PR DESCRIPTION
**Pull request by est31**
This PR does two things:

* Changes the default level of `pointer_structural_match` from allow to warn
* Turns `pointer_structural_match` into a normal lint instead of being a forward compatibility lint

The `pointer_structural_match` lint was added in #70743 to look for code that matches on a function pointer. As pointed out in #70861, and #54685, LLVM's behaviour makes function pointers behave in unexpected ways: function pointers might not be equal to "themselves" if function pointers for the same function were created in two different crates, and on the other hand, two different functions might end up with the same address if LLVM detects that they have the same body. See the linked issues for examples.

Thus, in #70743 an allow-by-default forward compatibility lint was added to address matching on pointers.

In https://github.com/rust-lang/rust/issues/70861#issuecomment-610065569 preferences were expressed that there should be a lint for matching on function pointers, but it should not be disallowed completely. I'd interpret that as wanting a warn-by-default non forward compatibility lint. Therefore, there are two possible ways forward:

* turn `pointer_structural_match` into a normal lint that's not a forward compatibility lint, and make it warn by default, OR
* keep it as a forward compatibility lint, to eventually forbid it via a compiler error, but first make it warn by default, then deny by default.

I think one should at least do one of the two things, even if it is only going to be a warning lint in the end.

The lang team has given the green light for the first approach. There is still the possibility that it will be forbidden in the future, but this will require concrete deprecation plans which don't exist right now.

cc @eddyb @pnkfelix @oli-obk

### CODE REVIEW

**Comment by est31** [comment]
cc also #74446 which contains some discussion of whether to allow fn pointers in match or not.

**Comment by oli-obk** [comment]
r? lang

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 1]
```diff
@@ -2285,7 +2285,7 @@ declare_lint! {
     /// to the same location. Additionally functions may get duplicated if they are instantiated
     /// in different crates and not deduplicated again via LTO.
     pub POINTER_STRUCTURAL_MATCH,
-    Allow,
+    Warn,
     "pointers are not structural-match",
     @future_incompatible = FutureIncompatibleInfo {
         reference: "issue #62411 <https://github.com/rust-lang/rust/issues/70861>",
```
Comment (tmandry) [inline compiler/rustc_lint_defs/src/builtin.rs:2291]: There are two related issues here, but in different formats. Are they supposed to refer to the same one?

Comment (tmandry) [reply compiler/rustc_lint_defs/src/builtin.rs:2291]:
```suggestion
        reference: "issue #70861 <https://github.com/rust-lang/rust/issues/70861>",
```

**Comment by nikomatsakis** [comment]
We discussed this in our lang team meeting today. The consensus was:

* We would like to warn about this.
* We would prefer not to issue a FCW unless we have a specific plan to remove it, which at the moment doesn't exist. But we would be open to the idea of creating such a plan in the future, and we can go from warning to FCW.

**Comment by est31** [comment]
I have changed the PR to now be warn by default and not be a forward compatibility lint.

@rustbot ready

**Comment by tmandry** [comment]
@rfcbot fcp merge

**Comment by nikomatsakis** [comment]
@rfcbot fcp reviewed

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 2]
```diff
@@ -2277,19 +2277,16 @@ declare_lint! {
...
-    /// optimizations pointers are "not equal to themselves" or pointers to different functions
-    /// compare as equal during runtime. This is because LLVM optimizations can deduplicate
-    /// functions if their bodies are the same, thus also making pointers to these functions point
-    /// to the same location. Additionally functions may get duplicated if they are instantiated
+    /// Use of function pointers and wide raw pointers in patterns works in many cases as
+    /// expected by users. However, it is possible that due to optimizations pointers are "not equal
+    /// to themselves" or pointers to different functions compare as equal during runtime.
+    /// This is because LLVM optimizations can deduplicate functions if their bodies are the same,
+    /// thus also making pointers to these functions point to the same location.
+    /// Additionally, functions may get duplicated if they are instantiated
```
Comment (RalfJung) [inline compiler/rustc_lint_defs/src/builtin.rs:2285]: Should probably also mention vtables
```suggestion
    /// Additionally, functions and vtables may get duplicated if they are instantiated
```

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 3]
```diff
@@ -2277,19 +2277,16 @@ declare_lint! {
     ///
     /// ### Explanation
     ///
-    /// Previous versions of Rust allowed function pointers and wide raw pointers in patterns.
-    /// While these work in many cases as expected by users, it is possible that due to
-    /// optimizations pointers are "not equal to themselves" or pointers to different functions
-    /// compare as equal during runtime. This is because LLVM optimizations can deduplicate
-    /// functions if their bodies are the same, thus also making pointers to these functions point
-    /// to the same location. Additionally functions may get duplicated if they are instantiated
+    /// Use of function pointers and wide raw pointers in patterns works in many cases as
```
Comment (RalfJung) [inline compiler/rustc_lint_defs/src/builtin.rs:2280]:
```suggestion
    /// Use of function pointers and `dyn Trait` raw pointers in patterns works in many cases as
```
I assume this doesn't fire on slice raw pointers?

**Comment by RalfJung** [comment]
@oli-obk IIRC this was a future-compat lint because of our plans for match and value trees. With the current state of those plans, do we still need the future compat part? I assume no since https://github.com/rust-lang/rust/pull/105750 landed. Function pointers and wide raw pointers do not have 'structural equality' but we allow them in `match` anyway because... backwards compatibility, I assume? In a perfect world we would only allow matching on things that have structural equality?

So I guess this means we have no plans to change that, and instead accept that we allow matching on some non-structural-eq types. We have a pretty clear semantics for that now (just use `PartialEq`) so that seems fine to me. But I wonder if the name and scope of the lint should be adjusted? The warning applies equally to `==` on function pointers and `dyn Trait` raw pointers.

**Comment by est31** [comment]
> With the current state of those plans, do we still need the future compat part? I assume no since https://github.com/rust-lang/rust/pull/105750 landed.

There are a bunch of forward compat lints, including the one for floats, that might be removed if that's the direction you want to go in. Note that this PR is already removing the future compat part, but only for `pointer_structural_match`.

> But I wonder if the name and scope of the lint should be adjusted? The warning applies equally to `==` on function pointers and `dyn Trait` raw pointers.

Hmm yeah good point, ideally we'd warn both for == and on matches, and then the name should probably be adjusted, too.

**Comment by RalfJung** [comment]
> There are a bunch of forward compat lints, including the one for floats, that might be removed if that's the direction you want to go in. Note that this PR is already removing the future compat part, but only for pointer_structural_match.

Ah, I forgot about those. Either we plan to eventually reject these non-structural-eq-matches, then they should all remain future-compat lints, or we don't plan to reject such code, then we probably should remove the future-compat status for all of them.

**Comment by est31** [comment]
So to summarize what @RalfJung suggested:

* I should make this lint also warn on `==` comparisons
* Probably a rename is needed... `fn_pointer_comparisons`, or, if `dyn Trait` is also affected, `unstable_address_comparisons` maybe?

**Comment by RalfJung** [comment]
@est31 what is the status of this? I recently dug into the entire pattern matching situation a bit more and I fully agree now that we don't want this to be a future-incompat lint any more. (And same for `illegal_floating_point_literal_pattern`.)

**Comment by est31** [comment]
I need to find time to update this PR with the listed todos. If someone else wants to beat me to it, please be my guest.

Regarding `illegal_floating_point_literal_pattern`, do you think it should just lose the forward compat status but remain warn-by default? I don't have any preference in either direction, but there needs to be a new reason/story for why the lint is there, and why we don't warn for `==` float comparisons (warning for such would be extremely disruptive by default).

**Comment by RalfJung** [comment]
Yeah that's the thing, the entire justification and wording of the lint was in the anticipation of fwd compat changes. If those are not going to happen we should start from scratch, i.e., remove the lint and think about what we actually want to lint against and why, and have a new lint for that. That can happen in the same PR but it doesn't have to.

If we don't lint against `==` on floats I see no good reason to lint against `match` on floats.

**Comment by Dylan-DPC** [comment]
@est31 any updates on this?

**Comment by RalfJung** [comment]
I think this is superseded by https://github.com/rust-lang/rust/pull/120423.

**Comment by est31** [comment]
Yeah let's close it.

**Pull request closed without merging**